File: projects/java/src/test/java/com/stater/statemachine/java/TestDoor.java
package com.stater.statemachine.java;

import org.junit.jupiter.api.Test;

import java.util.*;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;

public class TestDoor {
    private enum States {
        CLOSE, AJAR, OPEN
    }

    private static class DoorFSMContext implements Context {
        private Integer degreeOfOpening = 100;

        DoorFSMContext() {
        }

        DoorFSMContext(Integer degreeOfOpening) {
            this.degreeOfOpening = degreeOfOpening;
        }

        public Integer getDegreeOfOpening() {
            return degreeOfOpening;
        }

        public void setDegreeOfOpening(int degreeOfOpening) {
            this.degreeOfOpening = degreeOfOpening;
        }
    }

    private static class TypesDoorStateMachine extends StaterStateMachine<States, DoorFSMContext> {
        public TypesDoorStateMachine() {
            super(List.of(new Transition<>("preOpen", States.CLOSE, States.AJAR, ctx -> true, ctx -> ctx.setDegreeOfOpening(1)), new Transition<>("preClose", States.OPEN, States.AJAR, ctx -> true, ctx -> ctx.setDegreeOfOpening(99)), new Transition<>("open", States.AJAR, States.OPEN, ctx -> ctx.getDegreeOfOpening() >= 99, ctx -> ctx.setDegreeOfOpening(100)), new Transition<>("close", States.AJAR, States.CLOSE, ctx -> ctx.getDegreeOfOpening() <= 1, ctx -> ctx.setDegreeOfOpening(0)), new Transition<>("ajarPlus", States.AJAR, States.AJAR, ctx -> ctx.getDegreeOfOpening() >= 1 && ctx.getDegreeOfOpening() <= 98, ctx -> ctx.setDegreeOfOpening(ctx.getDegreeOfOpening() + 1)), new Transition<>("ajarMinus", States.AJAR, States.AJAR, ctx -> ctx.getDegreeOfOpening() >= 2 && ctx.getDegreeOfOpening() <= 99, ctx -> ctx.setDegreeOfOpening(ctx.getDegreeOfOpening() - 1))), new DoorFSMContext(), States.OPEN, Set.of(States.OPEN, States.CLOSE, States.AJAR), new HashMap<>(), new ArrayList<>(), new HashMap<>(), new ArrayList<>(), new HashMap<>(), new ArrayList<>(), null);
        }

        public TypesDoorStateMachine(List<Transition<States, DoorFSMContext>> transitions, DoorFSMContext context, States startState, Set<States> states, Map<String, List<TransitionMiddleware<DoorFSMContext>>> transitionMiddlewares, List<TransitionNameMiddleware<DoorFSMContext>> transitionAllMiddlewares, Map<String, List<Event<DoorFSMContext>>> transitionCallbacks, List<NameEvent<DoorFSMContext>> transitionAllCallbacks, Map<States, List<Event<DoorFSMContext>>> stateCallbacks, List<StateEvent<States, DoorFSMContext>> stateAllCallbacks, ContextJsonAdapter<DoorFSMContext> contextJsonAdapter) {
            super(transitions, context, startState, states, transitionMiddlewares, transitionAllMiddlewares, transitionCallbacks, transitionAllCallbacks, stateCallbacks, stateAllCallbacks, contextJsonAdapter);
        }

        public void ajarPlus() {
            this.transition("ajarPlus");
        }

        public void ajarMinus() {
            transition("ajarMinus");
        }

        public void preOpen() {
            transition("preOpen");
        }

        public void preClose() {
            transition("preClose");
        }

        public void open() {
            transition("open");
        }

        public void close() {
            transition("close");
        }
    }

    private final StateMachineFactory<States, DoorFSMContext> typedDoorFactory = TypesDoorStateMachine::new;


    private void testDoor(StaterStateMachine<States, DoorFSMContext> door) {
        assertEquals(door.getState(), States.OPEN);
        assertEquals(door.getContext().degreeOfOpening, 100);
        door.transition("preClose");
        assertEquals(door.getState(), States.AJAR);
        assertEquals(door.getContext().degreeOfOpening, 99);
        while (door.getContext().degreeOfOpening > 1) {
            door.transition("ajarMinus");
            assertEquals(door.getState(), States.AJAR);
        }
        assertEquals(door.getContext().degreeOfOpening, 1);
        door.transition("close");
        assertEquals(door.getContext().degreeOfOpening, 0);
        assertEquals(door.getState(), States.CLOSE);
        door.transition("preOpen");
        assertEquals(door.getContext().degreeOfOpening, 1);
        assertEquals(door.getState(), States.AJAR);
        door.transition("ajarPlus");
        assertEquals(door.getState(), States.AJAR);
        assertEquals(door.getContext().degreeOfOpening, 2);
        while (door.getContext().degreeOfOpening < 99) {
            door.transition("ajarPlus");
            assertEquals(door.getState(), States.AJAR);
        }
        door.transition("open");
        assertEquals(door.getState(), States.OPEN);
        assertEquals(door.getContext().degreeOfOpening, 100);
    }

    private void typedTestDoor(TypesDoorStateMachine door) {
        assertEquals(door.getState(), States.OPEN);
        assertEquals(door.getContext().degreeOfOpening, 100);
        door.preClose();
        assertEquals(door.getState(), States.AJAR);
        assertEquals(door.getContext().degreeOfOpening, 99);
        while (door.getContext().degreeOfOpening > 1) {
            door.ajarMinus();
            assertEquals(door.getState(), States.AJAR);
        }
        assertEquals(door.getContext().degreeOfOpening, 1);
        door.close();
        assertEquals(door.getContext().degreeOfOpening, 0);
        assertEquals(door.getState(), States.CLOSE);
        door.preOpen();
        assertEquals(door.getContext().degreeOfOpening, 1);
        assertEquals(door.getState(), States.AJAR);
        door.ajarPlus();
        assertEquals(door.getState(), States.AJAR);
        assertEquals(door.getContext().degreeOfOpening, 2);
        while (door.getContext().degreeOfOpening < 99) {
            door.ajarPlus();
            assertEquals(door.getState(), States.AJAR);
        }
        door.open();
        assertEquals(door.getState(), States.OPEN);
        assertEquals(door.getContext().degreeOfOpening, 100);
    }

    @Test
    public void testSimpleBuild() {
        final TypesDoorStateMachine doorFSM = new TypesDoorStateMachine();
        testDoor(doorFSM);
        typedTestDoor(doorFSM);
    }

    @Test
    public void testBuilder() {
        StaterStateMachine<States, DoorFSMContext> doorFSM = new StaterStateMachineBuilder<States, DoorFSMContext>()
                .addTransition("preOpen", States.CLOSE, States.AJAR, ctx -> true, context -> context.degreeOfOpening = 1)
                .addTransition("preClose", States.OPEN, States.AJAR, ctx -> true, context -> context.degreeOfOpening = 99)
                .addTransition("open", States.AJAR, States.OPEN, ctx -> ctx.degreeOfOpening >= 99, ctx -> ctx.degreeOfOpening = 100)
                .addTransition("close", States.AJAR, States.CLOSE, ctx -> ctx.degreeOfOpening <= 1, ctx -> ctx.degreeOfOpening = 0)
                .addTransition("ajarPlus", States.AJAR, States.AJAR, ctx -> 1 <= ctx.degreeOfOpening && ctx.degreeOfOpening <= 98, ctx -> ctx.degreeOfOpening++)
                .addTransition("ajarMinus", States.AJAR, States.AJAR, ctx -> 2 <= ctx.degreeOfOpening && ctx.degreeOfOpening <= 99, ctx -> ctx.degreeOfOpening--)
                .setContext(new DoorFSMContext())
                .setStartState(States.OPEN)
                .build();

        testDoor(doorFSM);
    }

    private StaterStateMachineBuilder<States, DoorFSMContext> structureBuild() {
        return new StaterStateMachineBuilder<States, DoorFSMContext>().addTransition("preOpen", States.CLOSE, States.AJAR).addTransition("preClose", States.OPEN, States.AJAR).addTransition("open", States.AJAR, States.OPEN).addTransition("close", States.AJAR, States.CLOSE).addTransition("ajarPlus", States.AJAR, States.AJAR).addTransition("ajarMinus", States.AJAR, States.AJAR);
    }

    private StaterStateMachineBuilder<States, DoorFSMContext> eventsBuild(StaterStateMachineBuilder<States, DoorFSMContext> builder) {
        return builder.setTransitionEvent("preOpen", ctx -> ctx.degreeOfOpening = 1).setTransitionEvent("preClose", ctx -> ctx.degreeOfOpening = 99).setTransitionCondition("open", ctx -> ctx.degreeOfOpening >= 99).setTransitionEvent("open", ctx -> ctx.degreeOfOpening = 100).setTransitionCondition("close", ctx -> ctx.degreeOfOpening <= 1).setTransitionEvent("close", ctx -> ctx.degreeOfOpening = 0).setTransitionCondition("ajarPlus", ctx -> 1 <= ctx.degreeOfOpening && ctx.degreeOfOpening <= 98).setTransitionEvent("ajarPlus", ctx -> ctx.degreeOfOpening++).setTransitionCondition("ajarMinus", ctx -> 2 <= ctx.degreeOfOpening && ctx.degreeOfOpening <= 99).setTransitionEvent("ajarMinus", ctx -> ctx.degreeOfOpening--);
    }

    @Test
    public void testBuilder2() {
        StaterStateMachine<States, DoorFSMContext> doorFSM = eventsBuild(structureBuild()).setContext(new DoorFSMContext()).setStartState(States.OPEN).build();

        testDoor(doorFSM);
    }

    @Test
    public void testAutoTransition() {
        StaterStateMachine<States, DoorFSMContext> doorFSM = eventsBuild(structureBuild()).setContext(new DoorFSMContext()).setStartState(States.OPEN).build();

        doorFSM.autoTransition();
        assertEquals(doorFSM.getState(), States.AJAR);
    }

    @Test
    public void testBuilderFactory() {
        StaterStateMachine<States, DoorFSMContext> doorFSM = eventsBuild(structureBuild()).setContext(new DoorFSMContext()).setStartState(States.OPEN).setFactory(typedDoorFactory).build();


        testDoor(doorFSM);
        assertInstanceOf(TypesDoorStateMachine.class, doorFSM);
        typedTestDoor((TypesDoorStateMachine) doorFSM);
    }

    @Test
    public void testJsonSchema() throws Exception {
        StaterStateMachine<States, DoorFSMContext> validDoorFSM = eventsBuild(structureBuild()).setContext(new DoorFSMContext()).setStartState(States.OPEN).build();

        testDoor(validDoorFSM);

        String jsonSchema = validDoorFSM.toJsonSchema();
        StaterStateMachine<States, DoorFSMContext> doorFSM = eventsBuild(new StaterStateMachineBuilder<States, DoorFSMContext>().fromJsonSchema(jsonSchema, States::valueOf)).setContext(new DoorFSMContext()).setFactory(typedDoorFactory).build();
        assertEquals(jsonSchema, doorFSM.toJsonSchema());

        testDoor(doorFSM);

        assertInstanceOf(TypesDoorStateMachine.class, doorFSM);
        typedTestDoor((TypesDoorStateMachine) doorFSM);
    }

    @Test
    public void testStringGeneric() throws Exception {
        StaterStateMachine<States, DoorFSMContext> validDoorFSM = eventsBuild(structureBuild()).setContext(new DoorFSMContext()).setStartState(States.OPEN).build();
        testDoor(validDoorFSM);
        String jsonSchema = validDoorFSM.toJsonSchema();
        StaterStateMachine<String, DoorFSMContext> stringDoorFSM = new StaterStateMachineBuilder<String, DoorFSMContext>().fromJsonSchema(jsonSchema, it -> it).setContext(new DoorFSMContext()).build();
        assertEquals(stringDoorFSM.getState(), "OPEN");
    }

    @Test
    public void testJsonDump() throws Exception {
        class JsonConverter implements ContextJsonAdapter<DoorFSMContext> {
            @Override
            public String toJson(DoorFSMContext context) {
                return context.degreeOfOpening.toString();
            }

            @Override
            public DoorFSMContext fromJson(String json) {
                return new DoorFSMContext(Integer.parseInt(json));
            }
        }

        StaterStateMachine<States, DoorFSMContext> validDoorFSM = eventsBuild(structureBuild()).setContext(new DoorFSMContext()).setStartState(States.OPEN).setContextJsonAdapter(new JsonConverter()).build();
        assertEquals(validDoorFSM.getState(), States.OPEN);
        assertEquals(validDoorFSM.getContext().degreeOfOpening, 100);
        String dump = validDoorFSM.toJson();
        validDoorFSM.transition("preClose");
        assertEquals(validDoorFSM.getState(), States.AJAR);
        assertEquals(validDoorFSM.getContext().degreeOfOpening, 99);
        validDoorFSM.fromJson(dump, States::valueOf);
        assertEquals(validDoorFSM.getState(), States.OPEN);
        assertEquals(validDoorFSM.getContext().degreeOfOpening, 100);
    }

    @Test
    public void testMiddlewareAndCallbacks() {
        AtomicReference<Integer> transitionMiddleware = new AtomicReference<>(0);
        AtomicReference<Integer> transitionAllMiddleware = new AtomicReference<>(0);
        AtomicReference<Integer> subscribeOnTransition = new AtomicReference<>(0);
        AtomicReference<Integer> subscribeOnAllTransition = new AtomicReference<>(0);
        AtomicReference<Integer> subscribeOnState = new AtomicReference<>(0);
        AtomicReference<Integer> subscribeOnAllState = new AtomicReference<>(0);

        StaterStateMachine<States, DoorFSMContext> validDoorFSM = eventsBuild(structureBuild()).setContext(new DoorFSMContext()).setStartState(States.OPEN).transitionMiddleware("open", (context, next) -> {
            transitionMiddleware.getAndSet(transitionMiddleware.get() + 1);
            next.accept(context);
        }).transitionAllMiddleware((name, context, next) -> {
            transitionAllMiddleware.getAndSet(transitionAllMiddleware.get() + 1);
            next.accept(name, context);
        }).subscribeOnTransition("open", a -> subscribeOnTransition.getAndSet(subscribeOnTransition.get() + 1)).subscribeOnAllTransition((a, b) -> subscribeOnAllTransition.getAndSet(subscribeOnAllTransition.get() + 1)).subscribeOnState(States.AJAR, a -> subscribeOnState.getAndSet(subscribeOnState.get() + 1)).subscribeOnAllState((a, b) -> subscribeOnAllState.getAndSet(subscribeOnAllState.get() + 1)).build();
        testDoor(validDoorFSM);
        assertEquals(transitionMiddleware.get(), 1);
        assertEquals(transitionAllMiddleware.get(), 200);
        assertEquals(subscribeOnTransition.get(), 1);
        assertEquals(subscribeOnAllTransition.get(), 200);
        assertEquals(subscribeOnState.get(), 198);
        assertEquals(subscribeOnAllState.get(), 200);
    }
}

File: projects/с#/Stater.StateMachine.Lib/Stater.StateMachine.Lib.Tests/TestDoor.cs
namespace Stater.StateMachine.Lib.Tests;

public class TestDoor
{
    private enum States
    {
        CLOSE,
        AJAR,
        OPEN
    }

    private class DoorFsmContext : IContext
    {
        public int DegreeOfOpening { get; set; } = 100;

        public DoorFsmContext()
        {
        }

        public DoorFsmContext(int degreeOfOpening)
        {
            DegreeOfOpening = degreeOfOpening;
        }

        public int GetDegreeOfOpening()
        {
            return DegreeOfOpening;
        }

        public void SetDegreeOfOpening(int degreeOfOpening)
        {
            DegreeOfOpening = degreeOfOpening;
        }
    }

    private class TypesDoorStateMachine : StaterStateMachine<States, DoorFsmContext>
    {
        public TypesDoorStateMachine()
            : base(
                new List<Transition<States, DoorFsmContext>>
                {
                    new("preOpen", States.CLOSE, States.AJAR, _ => true,
                        ctx => ctx.SetDegreeOfOpening(1)),
                    new("preClose", States.OPEN, States.AJAR, _ => true,
                        ctx => ctx.SetDegreeOfOpening(99)),
                    new("open", States.AJAR, States.OPEN,
                        ctx => ctx.GetDegreeOfOpening() >= 99, ctx => ctx.SetDegreeOfOpening(100)),
                    new("close", States.AJAR, States.CLOSE,
                        ctx => ctx.GetDegreeOfOpening() <= 1, ctx => ctx.SetDegreeOfOpening(0)),
                    new("ajarPlus", States.AJAR, States.AJAR,
                        ctx => ctx.GetDegreeOfOpening() >= 1 && ctx.GetDegreeOfOpening() <= 98,
                        ctx => ctx.SetDegreeOfOpening(ctx.GetDegreeOfOpening() + 1)),
                    new("ajarMinus", States.AJAR, States.AJAR,
                        ctx => ctx.GetDegreeOfOpening() >= 2 && ctx.GetDegreeOfOpening() <= 99,
                        ctx => ctx.SetDegreeOfOpening(ctx.GetDegreeOfOpening() - 1))
                },
                new DoorFsmContext(),
                States.OPEN,
                new HashSet<States> { States.OPEN, States.CLOSE, States.AJAR },
                new Dictionary<string, List<TransitionMiddleware<DoorFsmContext>>>(),
                new List<TransitionNameMiddleware<DoorFsmContext>>(),
                new Dictionary<string, List<Event<DoorFsmContext>>>(),
                new List<NameEvent<DoorFsmContext>>(),
                new Dictionary<States, List<Event<DoorFsmContext>>>(),
                new List<StateEvent<States, DoorFsmContext>>(),
                null
            )
        {
        }

        public TypesDoorStateMachine(
            List<Transition<States, DoorFsmContext>> transitions,
            DoorFsmContext context,
            States startState,
            HashSet<States> states,
            Dictionary<string, List<TransitionMiddleware<DoorFsmContext>>> transitionMiddlewares,
            List<TransitionNameMiddleware<DoorFsmContext>> transitionAllMiddlewares,
            Dictionary<string, List<Event<DoorFsmContext>>> transitionCallbacks,
            List<NameEvent<DoorFsmContext>> transitionAllCallbacks,
            Dictionary<States, List<Event<DoorFsmContext>>> stateCallbacks,
            List<StateEvent<States, DoorFsmContext>> stateAllCallbacks,
            IContextJsonAdapter<DoorFsmContext> contextJsonAdapter
        ) : base(
            transitions,
            context,
            startState,
            states,
            transitionMiddlewares,
            transitionAllMiddlewares,
            transitionCallbacks,
            transitionAllCallbacks,
            stateCallbacks,
            stateAllCallbacks,
            contextJsonAdapter
        )
        {
        }

        public void AjarPlus()
        {
            Transition("ajarPlus");
        }

        public void AjarMinus()
        {
            Transition("ajarMinus");
        }

        public void PreOpen()
        {
            Transition("preOpen");
        }

        public void PreClose()
        {
            Transition("preClose");
        }

        public void Open()
        {
            Transition("open");
        }

        public void Close()
        {
            Transition("close");
        }
    }


    private readonly StateMachineFactory<States, DoorFsmContext> _typedDoorFactory = (
        transitionsA,
        contextA,
        startStateA,
        statesA,
        transitionMiddlewaresA,
        transitionAllMiddlewaresA,
        transitionCallbacksA,
        transitionAllCallbacksA,
        stateCallbacksA,
        stateAllCallbacksA,
        contextJsonAdapterA
    ) => new TypesDoorStateMachine(
        transitionsA,
        contextA,
        startStateA,
        statesA,
        transitionMiddlewaresA,
        transitionAllMiddlewaresA,
        transitionCallbacksA,
        transitionAllCallbacksA,
        stateCallbacksA,
        stateAllCallbacksA,
        contextJsonAdapterA
    );

    private static void TestDoorSchema(StaterStateMachine<States, DoorFsmContext> door)
    {
        Assert.Equal(States.OPEN, door.GetState());
        Assert.Equal(100, door.GetContext().DegreeOfOpening);
        door.Transition("preClose");
        Assert.Equal(States.AJAR, door.GetState());
        Assert.Equal(99, door.GetContext().DegreeOfOpening);
        while (door.GetContext().DegreeOfOpening > 1)
        {
            door.Transition("ajarMinus");
            Assert.Equal(States.AJAR, door.GetState());
        }

        Assert.Equal(1, door.GetContext().DegreeOfOpening);
        door.Transition("close");
        Assert.Equal(0, door.GetContext().DegreeOfOpening);
        Assert.Equal(States.CLOSE, door.GetState());
        door.Transition("preOpen");
        Assert.Equal(1, door.GetContext().DegreeOfOpening);
        Assert.Equal(States.AJAR, door.GetState());
        door.Transition("ajarPlus");
        Assert.Equal(States.AJAR, door.GetState());
        Assert.Equal(2, door.GetContext().DegreeOfOpening);
        while (door.GetContext().DegreeOfOpening < 99)
        {
            door.Transition("ajarPlus");
            Assert.Equal(States.AJAR, door.GetState());
        }

        door.Transition("open");
        Assert.Equal(States.OPEN, door.GetState());
        Assert.Equal(100, door.GetContext().DegreeOfOpening);
    }

    private static void TypedTestDoor(TypesDoorStateMachine door)
    {
        Assert.Equal(States.OPEN, door.GetState());
        Assert.Equal(100, door.GetContext().DegreeOfOpening);
        door.PreClose();
        Assert.Equal(States.AJAR, door.GetState());
        Assert.Equal(99, door.GetContext().DegreeOfOpening);
        while (door.GetContext().DegreeOfOpening > 1)
        {
            door.AjarMinus();
            Assert.Equal(States.AJAR, door.GetState());
        }

        Assert.Equal(1, door.GetContext().DegreeOfOpening);
        door.Close();
        Assert.Equal(0, door.GetContext().DegreeOfOpening);
        Assert.Equal(States.CLOSE, door.GetState());
        door.PreOpen();
        Assert.Equal(1, door.GetContext().DegreeOfOpening);
        Assert.Equal(States.AJAR, door.GetState());
        door.AjarPlus();
        Assert.Equal(States.AJAR, door.GetState());
        Assert.Equal(2, door.GetContext().DegreeOfOpening);
        while (door.GetContext().DegreeOfOpening < 99)
        {
            door.AjarPlus();
            Assert.Equal(States.AJAR, door.GetState());
        }

        door.Open();
        Assert.Equal(States.OPEN, door.GetState());
        Assert.Equal(100, door.GetContext().DegreeOfOpening);
    }

    [Fact]
    public void TestSimpleBuild()
    {
        var doorFsm = new TypesDoorStateMachine();
        TestDoorSchema(doorFsm);
        TypedTestDoor(doorFsm);
    }

    [Fact]
    public void TestBuilder()
    {
        var doorFsm = new StaterStateMachineBuilder<States, DoorFsmContext>()
            .AddTransition("preOpen", States.CLOSE, States.AJAR, _ => true, context => context.DegreeOfOpening = 1)
            .AddTransition("preClose", States.OPEN, States.AJAR, _ => true, context => context.DegreeOfOpening = 99)
            .AddTransition("open", States.AJAR, States.OPEN, ctx => ctx.DegreeOfOpening >= 99,
                ctx => ctx.DegreeOfOpening = 100)
            .AddTransition("close", States.AJAR, States.CLOSE, ctx => ctx.DegreeOfOpening <= 1,
                ctx => ctx.DegreeOfOpening = 0)
            .AddTransition("ajarPlus", States.AJAR, States.AJAR,
                ctx => 1 <= ctx.DegreeOfOpening && ctx.DegreeOfOpening <= 98, ctx => ctx.DegreeOfOpening++)
            .AddTransition("ajarMinus", States.AJAR, States.AJAR,
                ctx => 2 <= ctx.DegreeOfOpening && ctx.DegreeOfOpening <= 99, ctx => ctx.DegreeOfOpening--)
            .SetContext(new DoorFsmContext())
            .SetStartState(States.OPEN)
            .Build();

        TestDoorSchema(doorFsm);
    }

    private static StaterStateMachineBuilder<States, DoorFsmContext> StructureBuild()
    {
        return new StaterStateMachineBuilder<States, DoorFsmContext>()
            .AddTransition("preOpen", States.CLOSE, States.AJAR).AddTransition("preClose", States.OPEN, States.AJAR)
            .AddTransition("open", States.AJAR, States.OPEN).AddTransition("close", States.AJAR, States.CLOSE)
            .AddTransition("ajarPlus", States.AJAR, States.AJAR).AddTransition("ajarMinus", States.AJAR, States.AJAR);
    }

    private static StaterStateMachineBuilder<States, DoorFsmContext> EventsBuild(
        StaterStateMachineBuilder<States, DoorFsmContext> builder)
    {
        return builder.SetTransitionEvent("preOpen", ctx => ctx.DegreeOfOpening = 1)
            .SetTransitionEvent("preClose", ctx => ctx.DegreeOfOpening = 99)
            .SetTransitionCondition("open", ctx => ctx.DegreeOfOpening >= 99)
            .SetTransitionEvent("open", ctx => ctx.DegreeOfOpening = 100)
            .SetTransitionCondition("close", ctx => ctx.DegreeOfOpening <= 1)
            .SetTransitionEvent("close", ctx => ctx.DegreeOfOpening = 0)
            .SetTransitionCondition("ajarPlus", ctx => 1 <= ctx.DegreeOfOpening && ctx.DegreeOfOpening <= 98)
            .SetTransitionEvent("ajarPlus", ctx => ctx.DegreeOfOpening++)
            .SetTransitionCondition("ajarMinus", ctx => 2 <= ctx.DegreeOfOpening && ctx.DegreeOfOpening <= 99)
            .SetTransitionEvent("ajarMinus", ctx => ctx.DegreeOfOpening--);
    }

    [Fact]
    public void TestBuilder2()
    {
        var doorFsm = EventsBuild(StructureBuild()).SetContext(new DoorFsmContext()).SetStartState(States.OPEN).Build();

        TestDoorSchema(doorFsm);
    }

    [Fact]
    public void TestAutoTransition()
    {
        var doorFsm = EventsBuild(StructureBuild()).SetContext(new DoorFsmContext()).SetStartState(States.OPEN).Build();

        doorFsm.AutoTransition();
        Assert.Equal(States.AJAR, doorFsm.GetState());
    }

    [Fact]
    public void TestBuilderFactory()
    {
        var doorFsm = EventsBuild(StructureBuild()).SetContext(new DoorFsmContext()).SetStartState(States.OPEN)
            .SetFactory(_typedDoorFactory).Build();


        TestDoorSchema(doorFsm);
        Assert.True(doorFsm is TypesDoorStateMachine);
        TypedTestDoor((TypesDoorStateMachine)doorFsm);
    }

    [Fact]
    public void TestJsonSchema()
    {
        var validDoorFsm = EventsBuild(StructureBuild()).SetContext(new DoorFsmContext()).SetStartState(States.OPEN)
            .Build();

        TestDoorSchema(validDoorFsm);

        var jsonSchema = validDoorFsm.ToJsonSchema();
        var doorFsm = EventsBuild(new StaterStateMachineBuilder<States, DoorFsmContext>()
                .FromJsonSchema(jsonSchema,
                    v =>
                    {
                        States value;
                        Enum.TryParse(v, out value);
                        return value;
                    }))
            .SetContext(new DoorFsmContext())
            .SetFactory(_typedDoorFactory)
            .Build();
        Assert.Equal(jsonSchema, doorFsm.ToJsonSchema());

        TestDoorSchema(doorFsm);

        Assert.True(doorFsm is TypesDoorStateMachine);
        TypedTestDoor((TypesDoorStateMachine)doorFsm);
    }

    [Fact]
    public void TestStringGeneric()
    {
        var validDoorFsm = EventsBuild(StructureBuild()).SetContext(new DoorFsmContext())
            .SetStartState(States.OPEN)
            .Build();
        TestDoorSchema(validDoorFsm);
        var jsonSchema = validDoorFsm.ToJsonSchema();
        var stringDoorFsm = new StaterStateMachineBuilder<string, DoorFsmContext>()
            .FromJsonSchema(jsonSchema, v =>
            {
                States value;
                Enum.TryParse(v, out value);
                return value.ToString();
            })
            .SetContext(new DoorFsmContext()).Build();
        Assert.Equal("OPEN", stringDoorFsm.GetState());
    }

    private class JsonConverter : IContextJsonAdapter<DoorFsmContext>
    {
        public string ToJson(DoorFsmContext context)
        {
            return context.DegreeOfOpening.ToString();
        }

        public DoorFsmContext FromJson(string json)
        {
            return new DoorFsmContext(int.Parse(json));
        }
    }

    [Fact]
    public void TestJsonDump()
    {
        StaterStateMachine<States, DoorFsmContext> validDoorFsm = EventsBuild(StructureBuild())
            .SetContext(new DoorFsmContext()).SetStartState(States.OPEN).SetContextJsonAdapter(new JsonConverter())
            .Build();
        Assert.Equal(States.OPEN, validDoorFsm.GetState());
        Assert.Equal(100, validDoorFsm.GetContext().DegreeOfOpening);
        var dump = validDoorFsm.ToJson();
        validDoorFsm.Transition("preClose");
        Assert.Equal(States.AJAR, validDoorFsm.GetState());
        Assert.Equal(99, validDoorFsm.GetContext().DegreeOfOpening);
        validDoorFsm.FromJson(dump, v =>
        {
            States value;
            Enum.TryParse(v, out value);
            return value;
        });
        Assert.Equal(States.OPEN, validDoorFsm.GetState());
        Assert.Equal(100, validDoorFsm.GetContext().DegreeOfOpening);
    }

    [Fact]
    public void TestMiddlewareAndCallbacks()
    {
        var transitionMiddleware = 0;
        var transitionAllMiddleware = 0;
        var subscribeOnTransition = 0;
        var subscribeOnAllTransition = 0;
        var subscribeOnState = 0;
        var subscribeOnAllState = 0;

        var validDoorFsm = EventsBuild(StructureBuild())
            .SetContext(new DoorFsmContext()).SetStartState(States.OPEN).TransitionMiddleware("open",
                (context, next) =>
                {
                    transitionMiddleware++;
                    next(context);
                })
            .TransitionAllMiddleware((name, context, next) =>
            {
                transitionAllMiddleware++;
                next(name, context);
            })
            .SubscribeOnTransition("open", _ => subscribeOnTransition++)
            .SubscribeOnAllTransition((_, _) => subscribeOnAllTransition++)
            .SubscribeOnState(States.AJAR, _ => subscribeOnState++)
            .SubscribeOnAllState((_, _) => subscribeOnAllState++)
            .Build();
        TestDoorSchema(validDoorFsm);
        Assert.Equal(1, transitionMiddleware);
        Assert.Equal(200, transitionAllMiddleware);
        Assert.Equal(1, subscribeOnTransition);
        Assert.Equal(200, subscribeOnAllTransition);
        Assert.Equal(198, subscribeOnState);
        Assert.Equal(200, subscribeOnAllState);
    }
}